File: scripts/test_huawei_auth.py
#!/usr/bin/env python3
"""华为终端认证测试脚本 - 兼容老版本TLS"""

import requests
import json
import ssl
import urllib3
from requests.adapters import HTTPAdapter
from urllib3.util.ssl_ import create_urllib3_context

# 禁用SSL警告
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

# 配置
BASE_URL = "https://10.62.10.3:443"
USERNAME = "api"
PASSWORD = "Hubei@1992"


class TLSAdapter(HTTPAdapter):
    """自定义TLS适配器，支持老版本的TLS配置"""

    def init_poolmanager(self, *args, **kwargs):
        # 创建支持老版本TLS的SSL上下文
        context = create_urllib3_context()

        # 设置TLS 1.2（华为TE40支持的版本）
        context.minimum_version = ssl.TLSVersion.TLSv1_2
        context.maximum_version = ssl.TLSVersion.TLSv1_2

        # 设置华为终端支持的密码套件
        # 使用OpenSSL格式的密码套件名称
        context.set_ciphers('HIGH:!aNULL:!eNULL:!MD5')

        # 不验证主机名（兼容自签名证书）
        context.check_hostname = False
        context.verify_mode = ssl.CERT_NONE

        kwargs['ssl_context'] = context
        return super().init_poolmanager(*args, **kwargs)


def mask_string(s):
    """隐藏敏感信息"""
    if len(s) <= 8:
        return "****"
    return s[:4] + "****" + s[-4:]


def test_huawei_auth():
    print("=== 华为终端认证流程测试 ===")
    print(f"目标: {BASE_URL}")
    print(f"用户名: {USERNAME}\n")

    # 创建session，使用自定义TLS适配器
    session = requests.Session()
    session.verify = False
    session.mount('https://', TLSAdapter())

    try:
        # 步骤1: 获取会话ID
        print("步骤1: 获取会话ID (Web_RequestSessionID)")
        session_url = f"{BASE_URL}/action.cgi?ActionID=Web_RequestSessionID"
        headers = {"userType": "web"}

        resp1 = session.post(session_url, headers=headers)
        print(f"  响应状态: {resp1.status_code}")
        print(f"  响应内容: {resp1.text}")

        data1 = resp1.json()
        if data1.get("success") != 1:
            print("  获取会话ID失败")
            return

        # 从Cookie中获取SessionID
        session_id = None
        for cookie in session.cookies:
            if cookie.name == "SessionID":
                session_id = cookie.value
                print(f"  从Cookie获取会话ID: {mask_string(session_id)}")
                break

        if not session_id:
            print("  错误: 未能获取到会话ID")
            return

        print(f"  最终会话ID: {mask_string(session_id)}\n")

        # 步骤2: 用户认证
        print("步骤2: 用户认证 (WEB_RequestCertificateAPI)")
        auth_url = f"{BASE_URL}/action.cgi?ActionID=WEB_RequestCertificateAPI"

        # 尝试不同的字段名格式
        auth_success = False
        for field_format in [("大写", "User", "Password"), ("小写", "user", "password")]:
            print(f"  尝试 {field_format[0]} 字段名:")

            auth_headers = {
                "userType": "web",
                "Content-Type": "application/json",
                "Cookie": f"SessionID={session_id}"
            }

            auth_payload = {
                field_format[1]: USERNAME,
                field_format[2]: PASSWORD
            }

            resp2 = session.post(auth_url, headers=auth_headers, json=auth_payload)
            print(f"    响应状态: {resp2.status_code}")
            print(f"    响应内容: {resp2.text[:200]}")

            data2 = resp2.json()
            if data2.get("success") == 1:
                print(f"    *** 认证成功! (使用{field_format[0]}字段名) ***\n")
                auth_success = True
                break
            else:
                error_id = data2.get("error", {}).get("id") or data2.get("exception", {}).get("id")
                print(f"    认证失败，错误码: {error_id}")

        print()

        # 步骤3: 替换会话ID
        if auth_success:
            print("步骤3: 替换会话ID (WEB_ChangeSessionID)")
            change_url = f"{BASE_URL}/action.cgi?ActionID=WEB_ChangeSessionID"

            change_headers = {
                "userType": "web",
                "Cookie": f"SessionID={session_id}"
            }

            resp3 = session.post(change_url, headers=change_headers)
            print(f"  响应状态: {resp3.status_code}")
            print(f"  响应内容: {resp3.text}")

            data3 = resp3.json()
            if data3.get("success") == 1:
                print("  *** 替换会话ID成功! ***")
            else:
                error_id = data3.get("error", {}).get("id")
                print(f"  替换失败，错误码: {error_id}")

    except Exception as e:
        print(f"错误: {e}")
        import traceback
        traceback.print_exc()

    print("\n=== 测试完成 ===")


if __name__ == "__main__":
    test_huawei_auth()

File: scripts/test_huawei_auth.go
package main

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type APIResponse struct {
	Success int    `json:"success"`
	Data    string `json:"data"`
	Error    struct {
		ID   int    `json:"id"`
		Code int    `json:"code"`
	} `json:"error,omitempty"`
}

type SessionIDResponse struct {
	AcSessionID string `json:"acSessionId"`
	SzTermType  string `json:"szTermType"`
}

func main() {
	baseURL := "https://10.62.10.3:443"
	username := "api"
	password := "Hubei@1992"

	fmt.Println("=== 华为终端认证流程测试 ===")
	fmt.Printf("目标: %s\n", baseURL)
	fmt.Printf("用户名: %s\n\n", username)

	// 创建HTTP客户端（跳过TLS验证，使用正确的密码套件）
	// 华为终端需要 TLS 1.0-1.2 版本范围和特定的密码套件
	client := &http.Client{
		Timeout: 10 * time.Second,
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{
				InsecureSkipVerify: true,
				MinVersion:         tls.VersionTLS10, // TLS 1.0
				MaxVersion:         tls.VersionTLS12, // 限制最大版本为 TLS 1.2
				// 华为终端支持的密码套件（尽可能多地包含以兼容老设备）
				CipherSuites: []uint16{
					tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
					tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
					tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
					tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
					tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
					tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
					tls.TLS_RSA_WITH_AES_256_CBC_SHA,
					tls.TLS_RSA_WITH_AES_128_CBC_SHA,
					tls.TLS_RSA_WITH_3DES_EDE_CBC_SHA,
				},
				// 使用支持所有密码套件的曲线偏好
				CurvePreferences: []tls.CurveID{
					tls.CurveP256,
					tls.CurveP384,
					tls.CurveP521,
				},
			},
		},
	}

	// 步骤1: 获取会话ID
	fmt.Println("步骤1: 获取会话ID (Web_RequestSessionID)")
	sessionURL := baseURL + "/action.cgi?ActionID=Web_RequestSessionID"
	req1, _ := http.NewRequest("POST", sessionURL, nil)
	req1.Header.Set("userType", "web")

	resp1, err := client.Do(req1)
	if err != nil {
		fmt.Printf("  请求失败: %v\n", err)
		return
	}
	defer resp1.Body.Close()

	body1, _ := io.ReadAll(resp1.Body)
	fmt.Printf("  响应状态: %d\n", resp1.StatusCode)
	fmt.Printf("  响应内容: %s\n", string(body1))

	// 解析响应
	var resp1Data APIResponse
	json.Unmarshal(body1, &resp1Data)

	if resp1Data.Success != 1 {
		fmt.Printf("  获取会话ID失败\n")
		return
	}

	// 从data字段解析会话ID
	var sessionResp SessionIDResponse
	json.Unmarshal([]byte(resp1Data.Data), &sessionResp)

	var sessionID string
	if sessionResp.AcSessionID != "" {
		sessionID = sessionResp.AcSessionID
		fmt.Printf("  从data获取会话ID: %s\n", maskString(sessionID))
	}

	// 从Cookie中提取会话ID
	if sessionID == "" {
		for _, cookie := range resp1.Cookies() {
			if cookie.Name == "SessionID" {
				sessionID = cookie.Value
				fmt.Printf("  从Cookie获取会话ID: %s\n", maskString(sessionID))
				break
			}
		}
	}

	if sessionID == "" {
		fmt.Printf("  错误: 未能获取到会话ID\n")
		return
	}

	fmt.Printf("  最终会话ID: %s\n", maskString(sessionID))
	fmt.Println()

	// 步骤2: 用户认证
	fmt.Println("步骤2: 用户认证 (WEB_RequestCertificateAPI)")
	authURL := baseURL + "/action.cgi?ActionID=WEB_RequestCertificateAPI"

	// 注意：华为API需要小写的user和password字段名
	authPayload := fmt.Sprintf(`{"user":"%s","password":"%s"}`, username, password)
	req2, _ := http.NewRequest("POST", authURL, strings.NewReader(authPayload))
	req2.Header.Set("Content-Type", "application/json")
	req2.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
	req2.Header.Set("userType", "web")
	req2.Header.Set("Cookie", fmt.Sprintf("SessionID=%s", sessionID))

	fmt.Printf("  请求Cookie: SessionID=%s\n", maskString(sessionID))

	resp2, err := client.Do(req2)
	if err != nil {
		fmt.Printf("  请求失败: %v\n", err)
		return
	}
	defer resp2.Body.Close()

	body2, _ := io.ReadAll(resp2.Body)
	fmt.Printf("  响应状态: %d\n", resp2.StatusCode)
	fmt.Printf("  响应内容: %s\n", string(body2))

	var resp2Data APIResponse
	json.Unmarshal(body2, &resp2Data)

	if resp2Data.Success == 1 {
		fmt.Printf("  *** 认证成功! ***\n")
	} else {
		fmt.Printf("  认证失败，错误码: %d\n", resp2Data.Error.ID)
	}
	fmt.Println()

	// 步骤3: 替换会话ID
	if resp2Data.Success == 1 {
		fmt.Println("步骤3: 替换会话ID (WEB_ChangeSessionID)")
		changeURL := baseURL + "/action.cgi?ActionID=WEB_ChangeSessionID"

		req3, _ := http.NewRequest("POST", changeURL, nil)
		req3.Header.Set("userType", "web")
		req3.Header.Set("Cookie", fmt.Sprintf("SessionID=%s", sessionID))

		resp3, err := client.Do(req3)
		if err != nil {
			fmt.Printf("  请求失败: %v\n", err)
			return
		}
		defer resp3.Body.Close()

		body3, _ := io.ReadAll(resp3.Body)
		fmt.Printf("  响应状态: %d\n", resp3.StatusCode)
		fmt.Printf("  响应内容: %s\n", string(body3))

		var resp3Data APIResponse
		json.Unmarshal(body3, &resp3Data)

		if resp3Data.Success == 1 {
			fmt.Printf("  *** 替换会话ID成功! ***\n")
		} else {
			fmt.Printf("  替换失败，错误码: %d\n", resp3Data.Error.ID)
		}
	}

	fmt.Println("\n=== 测试完成 ===")
}

// maskString 隐藏敏感信息
func maskString(s string) string {
	if len(s) <= 8 {
		return "****"
	}
	return s[:4] + "****" + s[len(s)-4:]
}
